File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l =0;
        int max =0;
    Set<Character> hi = new HashSet<>();
        for(int i=0;i< s.length();i++){
          char c = s.charAt(i);
          while(hi.contains(c)){
            hi.remove(s.charAt(l));
            l++;
          }
          hi.add(c);
          max = Math.max(max, i - l+1);
        }
        return max;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        freq = defaultdict(int)
        low = 0
        res = 0
        for high in range(len(s)):
            freq[s[high]] += 1
            while(len(freq) < (high - low + 1)):
                freq[s[low]] -= 1
                if(freq[s[low]] == 0):
                    del freq[s[low]]
                low += 1
            length = high - low  + 1
            res = max(res, length)
        return res
        